src/pipeline/error.rs

```diff
@@ -0,0 +1,32 @@
+use std::fmt::Debug;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed to decode message: {0}")]
+    DecodeError(#[from] prost::DecodeError),
+    #[error("failed to handle serde message: {0}")]
+    SerdeError(#[from] serde_json::Error),
+    #[error("failed to parse message: {0}")]
+    ParseError(#[from] std::str::Utf8Error),
+    #[error("problem communicating between threads: {0}")]
+    ThreadError(Box<dyn std::error::Error + Send + Sync>),
+}
+
+impl<T: Debug + Send + Sync + 'static> From<tokio::sync::mpsc::error::SendError<T>> for Error {
+    fn from(e: tokio::sync::mpsc::error::SendError<T>) -> Self {
+        Error::ThreadError(e.into())
+    }
+}
```

Comment:
for this particular `From`, the fact that the error has a generic `T` on it makes this the only sane way to implement, as far as I can tell.

For those below, I think `thiserror` strictly allows only a single `#[from]`

Reply:
and you're proposing adding 2 more entries to the enum? I liked the idea of bucketing the errors so we don't just have an explosion of types. Otherwise it feels like we've just recreated `Box<dyn ...>` but harder to use and called `Error`.

I have a macro that makes it easier to declare these (based on the macro that previously existed in this file):
```rust
macro_rules! multi_from {
    {$([$($t:path),+] => $v:ident),+ $(,)?} => {
        $($(
            impl From<$t> for Error {
                fn from(e: $t) -> Self {
                    Self::$v(e.into())
                }
            }
        )*)*
    }
}

multi_from! {
    [tokio::sync::oneshot::error::RecvError, tokio::task::JoinError] => ThreadError,
    [std::net::AddrParseError, std::str::Utf8Error] => ParseError,
}
```
so you don't need to define all those `impl From`s manually, if that helps (it may not...)